### DISCUSSION TITLE
pydantic v2 redoc and openapi bug

### DISCUSSION THREAD

**Discussion by InfernalAzazel** [fastapi/fastapi#9860] [Questions]
error: 

``` python
obj = Annotated[str, PlainValidator(lambda v: str(v))]
class Menu(BaseModel):
    uid: obj = Field(None, alias="_id", serialization_alias="_id")  # use Annotated error
```
effect: 

![image](https://github.com/tiangolo/fastapi/assets/40333954/0ecb761e-661b-41e2-a2b5-93e75041a13a)
![image](https://github.com/tiangolo/fastapi/assets/40333954/5898bac9-7f34-4704-8f9d-33491a41dd10)

Normal after annotation:

![image](https://github.com/tiangolo/fastapi/assets/40333954/2dde697e-1f05-4fdd-965d-7abd46ece83a)
![image](https://github.com/tiangolo/fastapi/assets/40333954/71645ba7-595c-45b7-abe1-5a4e6b267325)

There is also an issue with OpenAPI response_ The model did not display ALLMenuModel correctly, but it displayed normally in pydantic v1

![image](https://github.com/tiangolo/fastapi/assets/40333954/f408c0aa-36e4-4f38-9ca2-4b9741a9d752)

![image](https://github.com/tiangolo/fastapi/assets/40333954/cfd01917-88e9-42b6-a266-81ef525c2b03)

code:

- models
```python
from datetime import datetime
from typing import Any, List, Annotated, Optional

from pydantic import BaseModel, Field, PlainValidator

from app.utils.db import PyObjectId


class SearchMenu(BaseModel):
    key: int | None = None
    father: int | None = None
    path: str | None = None
    title_zh_cn: str | None = None
    title_en_us: str | None = None
    icon: str | None = None
    component: str | None = None
    order: int | None = None


class Menu(SearchMenu):
    uid: PyObjectId | None = Field(None, alias="_id")
    create_at: datetime | None = Field(default_factory=datetime.utcnow)  # 创建时间
    update_at: datetime | None = Field(default_factory=datetime.utcnow)  # 更新时间

    class Config:
        name = 'menu'


class MenuModel(Menu):
    children: list[Menu] = None


class ALLMenuModel(BaseModel):
    status_code: int
    success: bool
    detail: str = None
    data: List[MenuModel] | Any | None = None


class ListMenuModel(ALLMenuModel):
    total: int = 0
```

- app.utils.db.py

```python
from typing import Annotated
from zoneinfo import ZoneInfo

import motor.motor_asyncio
from pydantic import PlainValidator

from app.settings import MONGODB_USERNAME, MONGODB_PASSWORD, MONGODB_HOST, TZ_INFO, DATABASE_NAME

PyObjectId = Annotated[str, PlainValidator(lambda v: str(v))]


def async_db_engine():
    return motor.motor_asyncio.AsyncIOMotorClient(
        f'mongodb://{MONGODB_USERNAME}:{MONGODB_PASSWORD}@{MONGODB_HOST}',
        tz_aware=True,
        tzinfo=ZoneInfo(TZ_INFO)
    )[DATABASE_NAME]
```

Summary:

- Pydantic v2 is not fully adapted to redoc and openapi

version:

  • Installing pydantic-core (2.1.2)
  • Installing pydantic (2.0.2)
  • Installing starlette (0.27.0)
  • Installing fastapi (0.100.0)
  • Installing uvicorn (0.22.0)

**Comment by Kludex** [comment]
Please provide an MRE, and version of the packages used (FastAPI & Pydantic). 🙏 

Also... You sure? 👀 

<img width="459" alt="Screenshot 2023-07-12 at 11 00 42" src="https://github.com/tiangolo/fastapi/assets/7353520/845fd5a9-a928-45f3-b3ce-4c0b546a6cba">

**Reply by InfernalAzazel** [reply]
@tiangolo  It seems like there's nothing wrong with the logic? ....

**Reply by Kludex** [reply]
Please provide the imports 🙏

**Reply by InfernalAzazel** [reply]
A detailed version has been provided!

**Reply by InfernalAzazel** [reply]
The reproduction code is also provided. What do you need to provide to contact me
